sql/core/src/test/scala/org/apache/spark/sql/CollationSQLExpressionsSuite.scala

```diff
@@ -996,6 +996,30 @@ class CollationSQLExpressionsSuite
         assert(sql(query).schema.fields.head.dataType.sameType(dataType))
       }
     })
+
+    val tableName = "t_diff_collation"
+    withTable(tableName) {
+      sql(s"CREATE TABLE $tableName (" +
+        s"text STRING COLLATE UTF8_BINARY, " +
+        s"pairDelim STRING COLLATE UTF8_LCASE, " +
+        s"keyValueDelim STRING COLLATE UTF8_BINARY) " +
+        s"USING parquet")
+      checkError(
+        exception = intercept[AnalysisException] {
+          sql(s"SELECT str_to_map(text, pairDelim, keyValueDelim) from $tableName")
+        },
+        errorClass = "DATATYPE_MISMATCH.DATA_DIFF_TYPES",
+        sqlState = "42K09",
+        parameters = Map(
+          "functionName" -> "`str_to_map`",
+          "dataType" -> "[\"STRING\", \"STRING COLLATE UTF8_LCASE\", \"STRING\"]",
+          "sqlExpr" -> "\"str_to_map(text, pairDelim, keyValueDelim)\""),
+        context = ExpectedContext(
+          fragment = "str_to_map(text, pairDelim, keyValueDelim)",
+          start = 7,
+          stop = 48)
+      )
+    }
```

Comment:
we shouldn't use `DATATYPE_MISMATCH.DATA_DIFF_TYPES` for collation matching

please see `COLLATION_MISMATCH.EXPLICIT`, there should be many tests across the codebase market with "// Collation mismatch"